Project/Source/Components/ComponentParticleSystem.cpp

```diff
@@ -0,0 +1,325 @@
+		currentParticle.scale = float3(0.1f, 0.1f, 0.1f) * scale;
+		currentParticle.initialPosition = CreatePosition();
+		currentParticle.position = currentParticle.initialPosition;
+		currentParticle.direction = CreateVelocity();
+		currentParticle.velocity = vel;
+	}
+}
+
+void ComponentParticleSystem::Load(JsonValue jComponent) {
+	shaderID = jComponent[JSON_TAG_TEXTURE_SHADERID];
+
+	if (shaderID != 0) {
+		App->resources->IncreaseReferenceCount(shaderID);
+	}
+
+	textureID = jComponent[JSON_TAG_TEXTURE_TEXTUREID];
+
+	if (textureID != 0) {
+		App->resources->IncreaseReferenceCount(textureID);
+	}
+
+	CreateParticles(maxParticles, velocity);
+}
+
+void ComponentParticleSystem::Save(JsonValue jComponent) const {
+	jComponent[JSON_TAG_TEXTURE_SHADERID] = shaderID;
+	jComponent[JSON_TAG_TEXTURE_TEXTUREID] = textureID;
+}
+
+void ComponentParticleSystem::Update() {
+	deadParticles.clear();
+	for (Particle& currentParticle : particles) {
+		if (App->time->IsGameRunning()) {
+			currentParticle.life -= App->time->GetDeltaTime();
+		} else {
+			currentParticle.life -= App->time->GetRealTimeDeltaTime();
+		}
+		currentParticle.position += currentParticle.direction * velocity;
+		currentParticle.model = float4x4::FromTRS(currentParticle.position, currentParticle.rotation, currentParticle.scale);
+		if (currentParticle.life < 0) {
+			deadParticles.push_back(&currentParticle);
+		}
+	}
+	for (Particle* currentParticle : deadParticles) {
+		particles.Release(currentParticle);
+	}
+	Draw();
+}
+
+void ComponentParticleSystem::Init() {
+	CreateParticles(maxParticles, velocity);
+}
+
+void ComponentParticleSystem::SpawnParticle() {
+	Particle* currentParticle = particles.Obtain();
+	if (currentParticle) {
+		currentParticle->position = currentParticle->initialPosition;
+		currentParticle->life = particleLife;
+		if (isRandomFrame) {
+			currentParticle->currentFrame = rand() % ((Xtiles * Ytiles) + 1);
+		} else {
+			currentParticle->currentFrame = 0;
+		}
+		currentParticle->colorFrame = 0;
+		currentParticle->direction = CreateVelocity();
+		currentParticle->initialPosition = CreatePosition();
+		currentParticle->position = currentParticle->initialPosition;
+		//TODO: not hardcoded
+		currentParticle->scale = float3(0.1f, 0.1f, 0.1f) * scale;
+	}
+}
+
+void ComponentParticleSystem::killParticles() {
+	for (Particle& currentParticle : particles) {
+		particles.Release(&currentParticle);
+	}
+}
+
+void ComponentParticleSystem::DrawGizmos() {
+	//TODO: IMPROVE DRAWS
+	if (IsActive()) {
+		if (emitterType == EmitterType::CONE) {
+			ComponentTransform* transform = GetOwner().GetComponent<ComponentTransform>();
+			dd::cone(transform->GetGlobalPosition(), transform->GetGlobalRotation() * float3::unitY * 1, dd::colors::White, 1.0f, 0.3f);
+		}
+		if (emitterType == EmitterType::SPHERE) {
+			float delta = kl * kl - 4 * (kc - 10) * kq;
+			float distance = Max(abs((-kl + sqrt(delta))) / (2 * kq), abs((-kl - sqrt(delta)) / (2 * kq)));
```

Comment:
Already marked as TODO, but shouldnt it be a switch? With this if structure the program is testing a lot of Jumps